### PR TITLE
Select only the columns from the `wp_bp_notifications` table

### PR DESCRIPTION
**Pull request by renatonascalves**
When doing a notification request with the `meta_query` param, with the `NOT EXISTS` value, one can get the `id` columns from the `bp_notifications` and `bp_notifications_meta` tables. See the example below.

This query:
```php
BP_Notifications_Notification::get(
	[
		'user_id'    => 5,
		'meta_query' => [
			[
				'key'     => 'foo',
				'compare' => 'NOT EXISTS'
			]
		],
	]
);
```
Will create this mysql query:
```mysql
SELECT
	*
FROM
	wp_bp_notifications n
	LEFT JOIN wp_bp_notifications_meta ON (n.id = wp_bp_notifications_meta.notification_id
			AND wp_bp_notifications_meta.meta_key = 'foo')
WHERE
	user_id IN(5)
	AND component_name IN('activity', 'blogs', 'friends', 'groups', 'members', 'messages', 'xprofile')
	AND is_new = 1
	AND(wp_bp_notifications_meta.notification_id IS NULL)
ORDER BY
	date_notified
```

The result ends up with two `id` columns. WordPress is returning the one with `NULL` values.

Screenshot 2022-12-29 at 15 21 09

This pull request adds a fix for that, plus unit tests.

Trac ticket: https://buddypress.trac.wordpress.org/ticket/8556

---
**This Pull Request is for code review only. Please keep all other discussion in the BuddyPress Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the WordPress Core Handbook for more details.**

### CODE REVIEW

**Review by imath — CHANGES_REQUESTED** [review]
Thanks a lot for your work on this. I believe some changes are needed:

1. changes about `src/bp-activity/classes/class-bp-activity-activity.php` + the corresponding Unit Test are not concerning the ticket and consists of a better way to write the same code (returning early if a condition is not matched instead of only doing something if the condition matches). So I think these changes should be remove from this PR.
2. Problem with using `n.*` is that you are not getting fields from the metadata table, and the duplicate ID is not reflecting the same information: the first one is about the notification and the second one is about the metadata of this notification. I believe, all fields (except the one about the `notification_id`) from the metadata table should remain and the `id` field should be aliased as `meta_id` for instance. I think having the meta_key and meta_value fields can be important for these kind of queries. I'd look how iWordPress is handling this, if I were you. But ultimately we could have `{$metatable_name} as nm` and have a SELECT looking like this `SELECT n.*,nm.meta_key, nm.meta_value`.

**Comment by renatonascalves** [comment]
> changes about src/bp-activity/classes/class-bp-activity-activity.php + the corresponding Unit Test are not concerning the ticket and consist of a better way to write the same code (returning early if a condition is not matched instead of only doing something if the condition matches). So I think these changes should be removed from this PR.

1 - I'll keep the unit test since **IT IS** related to the task at hand (making sure the same issue is not present in the Activity component, hence the unit test). The other change was so minor that I don't see why we need to remove it out. But I'm happy to remove it if you feel strongly. I'm just not that _orthodox_ when it comes to small changes here and there ¯\_(ツ)_/¯

2 - @imath  🤔  I think this needs a bit more discussion due to its very narrow use case. Currently:

- The default request does not return data from the `bp_notifications_meta` table;
- No data is being used inside BuddyPress from the `bp_notifications_meta` table when using that request;
- The data is only available when using the `meta_query` param using the `EXISTS` value. 
- When using the `NOT EXISTS` value**, the data is `null`: there is no `id` (meta id), `notification_id`, `meta_key`, `meta_value`;

So it seems your suggestion is to:

- refactor the `::get` method to return this data only when using the `meta_query` param using the `EXISTS` value.

Confirm?!

File: src/bp-activity/classes/class-bp-activity-activity.php [hunk 1]
```diff
@@ -1049,31 +1049,38 @@ protected static function generate_action_strings( $activities ) {
 	 *
 	 * @since 1.8.0
 	 *
+	 * @global BuddyPress $bp The one true BuddyPress instance.
```
[the remark this answers is not in the document]
Comment (renatonascalves) [reply src/bp-activity/classes/class-bp-activity-activity.php:1052]: Noded!

File: src/bp-notifications/classes/class-bp-notifications-notification.php [hunk 2]
```diff
@@ -813,11 +812,15 @@ public static function get_total_count( $args ) {
 	 *
 	 * @since 2.3.0
 	 *
+	 * @global BuddyPress $bp The one true BuddyPress instance.
```
[the remark this answers is not in the document]
Comment (renatonascalves) [reply src/bp-notifications/classes/class-bp-notifications-notification.php:815]: Noded!

**Comment by renatonascalves** [comment]
By the way, I'm happy to commit number 1 in a separate SVN commit since that's your main contention.

**Comment by imath** [comment]
I confirm 🙂.

**Review by imath — COMMENTED** [review]
Perfect! Thanks a lot @renatonascalves. Green lights for SVN trunk 😉

**Pull request closed without merging**